File: tws-graph/src/tws_graph/indexer/extractors/toml_extractor.py
"""TOML extractor — standalone (does not inherit BaseExtractor, per P6 freeze).

Produces CONTAINS edges for:
- Top-level tables: [name]
- Dotted tables: [parent.child]
- Array-of-tables: [[name]]
- Key-value pairs: key = value

Usage:
    from tree_sitter_language_pack import get_parser
    parser = get_parser("toml")
    tree = parser.parse(source)
    edges = extract(source.encode(), tree, "config.toml")
"""

from __future__ import annotations

from tws_graph.indexer.base import hash_id, BaseExtractor


def _node_text(node, source: bytes) -> str:
    return source[node.start_byte():node.end_byte()].decode("utf-8")


def _named_children(node):
    for i in range(node.named_child_count()):
        yield node.named_child(i)


def _children(node):
    for i in range(node.child_count()):
        yield node.child(i)


def _make_edge(source_id: str, target_text: str, kind: str, file_path: str, line: int) -> dict:
    return {
        "source": source_id,
        "target": "",
        "kind": kind,
        "target_text": target_text,
        "source_loc": f"{file_path}:{line}",
        "provenance": "tree-sitter",
    }


def _resolve_dotted_key(node, source: bytes) -> str:
    """Resolve a dotted_key or bare_key node to a dotted name string."""
    if node.kind() == "bare_key":
        return _node_text(node, source)
    if node.kind() == "dotted_key":
        parts = []
        for child in _children(node):
            if child.is_named and child.kind() in ("bare_key", "dotted_key"):
                parts.append(_resolve_dotted_key(child, source))
        return ".".join(parts)
    return ""


def _resolve_pair_text(node, source: bytes) -> str:
    """Resolve a pair node to 'key=value' text."""
    key = ""
    value = ""
    for child in _named_children(node):
        if child.kind() in ("bare_key", "dotted_key", "quoted_key"):
            if child.kind() == "quoted_key":
                key = _node_text(child, source)
            else:
                key = _resolve_dotted_key(child, source)
        elif key and child.kind() not in ("=",) and child.is_named:
            value = _node_text(child, source)
            break
    if key and value:
        return f"{key}={value}"
    return ""


def extract(source: bytes, tree, file_path: str) -> list[dict]:
    """Extract CONTAINS edges from a TOML CST.

    Args:
        source: Raw file bytes.
        tree: tree-sitter parse result.
        file_path: Logical file path (used in source IDs and locs).

    Returns:
        List of edge dicts.
    """
    edges: list[dict] = []
    root = tree.root_node()

    def walk(node, current_section: str = ""):
        for child in _children(node):
            kind = child.kind()

            if kind == "table":
                # [name] or [dotted.key]
                key_node = None
                for c in _named_children(child):
                    if c.kind() in ("bare_key", "dotted_key"):
                        key_node = c
                        break
                if key_node:
                    section = _resolve_dotted_key(key_node, source)
                    source_id = hash_id(f"{file_path}::{section}", file_path)
                    edges.append(_make_edge(
                        source_id, section, "contains",
                        file_path, child.start_position().row + 1,
                    ))
                    # Recurse with new section context
                    for c in _named_children(child):
                        if c.kind() == "pair":
                            _handle_pair(c, source_id, file_path)
                    # continue walking into pairs
                    continue

            elif kind == "table_array_element":
                # [[name]]
                key_node = None
                for c in _named_children(child):
                    if c.kind() in ("bare_key", "dotted_key"):
                        key_node = c
                        break
                if key_node:
                    section = _resolve_dotted_key(key_node, source)
                    source_id = hash_id(f"{file_path}::{section}", file_path)
                    edges.append(_make_edge(
                        source_id, section, "contains",
                        file_path, child.start_position().row + 1,
                    ))
                    for c in _named_children(child):
                        if c.kind() == "pair":
                            _handle_pair(c, source_id, file_path)
                    continue

            # walk children recursively
            walk(child, current_section)

    def _handle_pair(pair_node, parent_id: str, file_path: str):
        """Handle a key=value pair."""
        text = _resolve_pair_text(pair_node, source)
        if text:
            edges.append(_make_edge(
                parent_id, text, "contains",
                file_path, pair_node.start_position().row + 1,
            ))

    walk(root)
    return edges


class TomlExtractor(BaseExtractor):
    extensions = [".toml"]
    tree_sitter_languages = ["toml"]

    def extract(self, source, tree, ctx) -> None:
        result_edges = extract(source, tree, ctx.file_path)
        ctx.result.edges.extend(result_edges)

File: tws-graph/rust_core/src/indexer/extractors/toml_extractor.rs
//! TOML language extractor.
//!
//! Extracts symbols and relationships from TOML source files (`.toml`)
//! using the tree-sitter-toml-ng grammar.
//!
//! # Node kinds produced
//! - `toml_table`: TOML table headers `[table]`
//! - `toml_table_array`: TOML array of tables `[[array]]`
//!
//! # Edge kinds produced
//! - `contains`: containment (parent table -> child table and key-value pairs)

use crate::indexer::context::ExtractionContext;
use crate::traits::{EdgeKind, Extractor, NodeKind};
use std::collections::HashMap;
use tree_sitter::Node;
use tree_sitter::Tree;

pub struct TomlExtractor;

impl Extractor for TomlExtractor {
    fn extensions(&self) -> Vec<&'static str> {
        vec!["toml"]
    }
    fn languages(&self) -> Vec<&'static str> {
        vec!["toml"]
    }
    fn extract(
        &self,
        source: &[u8],
        tree: &Tree,
        ctx: &mut ExtractionContext,
    ) -> anyhow::Result<()> {
        let root = tree.root_node();
        let file_id = ctx.add_node(NodeKind::File, "toml", &root, HashMap::new());

        walk_toml(source, root, ctx, &file_id)?;

        Ok(())
    }
}

/// Find a direct named child by kind.
fn find_child<'a>(node: Node<'a>, kind: &str) -> Option<Node<'a>> {
    for i in 0..node.named_child_count() {
        if let Some(child) = node.named_child(i) {
            if child.kind() == kind {
                return Some(child);
            }
        }
    }
    None
}

/// Get the key name from a table or table_array_element header.
/// In toml-ng grammar: table → [, bare_key/dotted_key, ], pair...
fn get_header_name(source: &[u8], node: Node) -> String {
    // Check for bare_key first
    if let Some(key_node) = find_child(node, "bare_key") {
        return get_text(source, Some(key_node));
    }
    // Check for dotted_key
    if let Some(key_node) = find_child(node, "dotted_key") {
        return get_text(source, Some(key_node));
    }
    // Fallback: iterate named children for key-like nodes
    for i in 0..node.named_child_count() {
        if let Some(child) = node.named_child(i) {
            let kind = child.kind();
            if kind == "bare_key" || kind == "dotted_key" {
                return get_text(source, Some(child));
            }
        }
    }
    String::new()
}

fn walk_toml(
    source: &[u8],
    node: Node,
    ctx: &mut ExtractionContext,
    parent_id: &str,
) -> anyhow::Result<()> {
    for i in 0..node.named_child_count() {
        if let Some(child) = node.named_child(i) {
            match child.kind() {
                "table" => {
                    extract_table(source, child, ctx, parent_id)?;
                }
                "table_array_element" => {
                    extract_table_array(source, child, ctx, parent_id)?;
                }
                "pair" => {
                    extract_pair(source, child, ctx, parent_id)?;
                }
                _ => {
                    walk_toml(source, child, ctx, parent_id)?;
                }
            }
        }
    }
    Ok(())
}

fn extract_table(
    source: &[u8],
    node: Node,
    ctx: &mut ExtractionContext,
    parent_id: &str,
) -> anyhow::Result<(String, usize)> {
    let header_text = get_header_name(source, node);
    if header_text.is_empty() {
        return Ok((String::new(), 0));
    }

    let line = node.start_position().row as u32 + 1;
    let table_id = ctx.add_node(NodeKind::TomlTable, &header_text, &node, HashMap::new());
    ctx.add_edge(parent_id, &table_id, EdgeKind::Contains, line, None);

    let mut pair_count = 0;
    // Walk children to find pairs and nested tables
    for i in 0..node.named_child_count() {
        if let Some(child) = node.named_child(i) {
            match child.kind() {
                "pair" => {
                    extract_pair(source, child, ctx, &table_id)?;
                    pair_count += 1;
                }
                "table" => {
                    extract_table(source, child, ctx, &table_id)?;
                }
                "table_array_element" => {
                    extract_table_array(source, child, ctx, parent_id)?;
                }
                _ => {}
            }
        }
    }

    Ok((header_text, pair_count))
}

fn extract_table_array(
    source: &[u8],
    node: Node,
    ctx: &mut ExtractionContext,
    parent_id: &str,
) -> anyhow::Result<()> {
    let header_text = get_header_name(source, node);
    if header_text.is_empty() {
        return Ok(());
    }

    let line = node.start_position().row as u32 + 1;
    let ta_id = ctx.add_node(NodeKind::TomlTableArray, &header_text, &node, HashMap::new());
    ctx.add_edge(parent_id, &ta_id, EdgeKind::Contains, line, None);

    // Walk children for pairs
    for i in 0..node.named_child_count() {
        if let Some(child) = node.named_child(i) {
            if child.kind() == "pair" {
                extract_pair(source, child, ctx, &ta_id)?;
            }
        }
    }

    Ok(())
}

fn extract_pair(
    source: &[u8],
    node: Node,
    ctx: &mut ExtractionContext,
    parent_id: &str,
) -> anyhow::Result<()> {
    // In toml-ng grammar: pair → key_node = value_node
    // Key is the first named child (bare_key, dotted_key, quoted_key, etc.)
    let key_text = if let Some(key_node) = node.named_child(0) {
        get_text(source, Some(key_node))
    } else {
        String::new()
    };

    if key_text.is_empty() {
        return Ok(());
    }

    let line = node.start_position().row as u32 + 1;
    let key_id = ctx.add_node(NodeKind::Field, &key_text, &node, HashMap::new());
    ctx.add_edge(parent_id, &key_id, EdgeKind::Contains, line, None);

    Ok(())
}

fn get_text(source: &[u8], node: Option<Node>) -> String {
    match node {
        Some(n) => n
            .utf8_text(source)
            .map(|c| c.to_string())
            .unwrap_or_default(),
        None => String::new(),
    }
}

// ---------------------------------------------------------------------------
// Tests
// ---------------------------------------------------------------------------

#[cfg(test)]
mod tests {
    use super::*;
    use crate::indexer::context::ExtractionContext;
    use crate::traits::{EdgeKind, NodeKind};
    use tree_sitter::Parser;

    fn extract(source: &str, file_path: &str) -> ExtractionContext {
        let mut parser = Parser::new();
        parser
            .set_language(&tree_sitter_toml_ng::LANGUAGE.into())
            .expect("set toml language");
        let tree = parser.parse(source, None).expect("parse toml source");

        let mut ctx = ExtractionContext::new(file_path.to_string(), "toml".to_string());
        TomlExtractor
            .extract(source.as_bytes(), &tree, &mut ctx)
            .expect("extract should succeed");
        ctx
    }

    fn find_nodes<'a>(ctx: &'a ExtractionContext, kind: NodeKind) -> Vec<&'a crate::db::models::NodeRecord> {
        let kind_str = crate::indexer::context::node_kind_to_str(kind);
        ctx.result.nodes.iter().filter(|n| n.kind == kind_str).collect()
    }

    fn find_edges<'a>(ctx: &'a ExtractionContext, kind: EdgeKind) -> Vec<&'a crate::db::models::EdgeRecord> {
        let kind_str = kind.as_str();
        ctx.result.edges.iter().filter(|e| e.kind == kind_str).collect()
    }

    #[test]
    fn test_extract_simple_table() {
        let ctx = extract("[package]\nname = \"myapp\"\n", "src/test.toml");
        let tables = find_nodes(&ctx, NodeKind::TomlTable);
        assert_eq!(tables.len(), 1, "Expected 1 toml_table, got {:?}", tables.iter().map(|t| t.name.as_str()).collect::<Vec<_>>());
        assert_eq!(tables[0].name, "package");
    }

    #[test]
    fn test_extract_dotted_table() {
        let ctx = extract("[package.metadata]\nkey = \"value\"\n", "src/test.toml");
        let tables = find_nodes(&ctx, NodeKind::TomlTable);
        assert_eq!(tables.len(), 1, "Expected 1 table, got {:?}", tables.iter().map(|t| t.name.as_str()).collect::<Vec<_>>());
        assert_eq!(tables[0].name, "package.metadata");
    }

    #[test]
    fn test_extract_multiple_tables() {
        let ctx = extract(
            "[package]\nname = \"app\"\n\n[dependencies]\nlib = \"1.0\"\n",
            "src/test.toml",
        );
        let tables = find_nodes(&ctx, NodeKind::TomlTable);
        assert_eq!(tables.len(), 2, "Expected 2 tables, got {:?}", tables.iter().map(|t| t.name.as_str()).collect::<Vec<_>>());
        let names: Vec<&str> = tables.iter().map(|n| n.name.as_str()).collect();
        assert!(names.contains(&"package"));
        assert!(names.contains(&"dependencies"));
    }

    #[test]
    fn test_extract_table_array() {
        let ctx = extract("[[servers]]\nhost = \"a\"\n\n[[servers]]\nhost = \"b\"\n", "src/test.toml");
        let ta = find_nodes(&ctx, NodeKind::TomlTableArray);
        assert_eq!(ta.len(), 2, "Expected 2 table_arrays, got {:?}", ta.iter().map(|t| t.name.as_str()).collect::<Vec<_>>());
        assert!(ta[0].name.contains("servers"));
        assert!(ta[1].name.contains("servers"));
    }

    #[test]
    fn test_extract_key_value_pairs() {
        let ctx = extract("[package]\nname = \"myapp\"\nversion = \"1.0\"\n", "src/test.toml");
        let fields = find_nodes(&ctx, NodeKind::Field);
        assert_eq!(fields.len(), 2, "Expected 2 fields, got {:?}", fields.iter().map(|f| f.name.as_str()).collect::<Vec<_>>());
        let names: Vec<&str> = fields.iter().map(|n| n.name.as_str()).collect();
        assert!(names.contains(&"name"));
        assert!(names.contains(&"version"));
    }

    #[test]
    fn test_extract_empty_file() {
        let ctx = extract("", "src/empty.toml");
        let files = find_nodes(&ctx, NodeKind::File);
        assert_eq!(files.len(), 1);
    }

    #[test]
    fn test_extract_root_keys() {
        let ctx = extract("title = \"My Config\"\ndescription = \"A test config\"\n", "src/test.toml");
        let fields = find_nodes(&ctx, NodeKind::Field);
        assert_eq!(fields.len(), 2, "Expected 2 fields for root keys, got {:?}", fields.iter().map(|f| f.name.as_str()).collect::<Vec<_>>());
    }

    #[test]
    fn test_extract_deeply_nested_table() {
        let ctx = extract("[a.b.c.d]\nkey = \"value\"\n", "src/test.toml");
        let tables = find_nodes(&ctx, NodeKind::TomlTable);
        assert_eq!(tables.len(), 1, "Expected 1 table, got {:?}", tables.iter().map(|t| t.name.as_str()).collect::<Vec<_>>());
        // dotted_key creates the name as "a.b.c.d"
        assert_eq!(tables[0].name, "a.b.c.d", "Expected dotted name a.b.c.d, got '{}'", tables[0].name);
    }

    #[test]
    fn test_extract_contains_edges() {
        let ctx = extract("[package]\nname = \"app\"\n", "src/test.toml");
        let contains = find_edges(&ctx, EdgeKind::Contains);
        assert!(contains.len() >= 2, "Expected at least 2 contains edges (file->table, table->field), got {:?}", contains.iter().map(|e| format!("{} -> {}", e.source, e.target)).collect::<Vec<_>>());
    }

    #[test]
    fn test_extract_complex_toml() {
        let ctx = extract(
            "[package]\nname = \"app\"\nversion = \"1.0\"\n\n[dependencies]\nserde = \"1.0\"\n\n[[bin]]\nname = \"cli\"\npath = \"src/main.rs\"\n",
            "src/test.toml",
        );
        let tables = find_nodes(&ctx, NodeKind::TomlTable);
        assert!(!tables.is_empty(), "Should have tables, got {:?}", ctx.result.nodes.iter().map(|n| format!("{}:{}", n.kind, n.name)).collect::<Vec<_>>());

        let ta = find_nodes(&ctx, NodeKind::TomlTableArray);
        assert!(!ta.is_empty(), "Should have table arrays");

        let fields = find_nodes(&ctx, NodeKind::Field);
        assert!(fields.len() >= 4, "Expected at least 4 field nodes, got {}", fields.len());
    }
}
